docs/v1/P4Runtime-Spec.mdk

```diff
@@ -4544,7 +4564,10 @@ four replicas of an ARP packet. These replicas will appear in the egress
 pipeline as independent packets with egress port set to PSA device port numbers
 corresponding to SDN port numbers 5, 12, 18 and 24. For more discussion on the
 translation between SDN ports and PSA device ports, refer to the
-[PSA Metadata Translation](#sec-translation-of-port-numbers) section.
+[PSA Metadata Translation](#sec-translation-of-port-numbers) section. Each
+replica can optionally have a list of backup replicas used as fallback ports
+for multicast. When the primary port (replica) becomes unavailable, the system
+uses the first available backup replica.
```

Comment:
Suppose one of the replicas for a multicast group is for port 5, instance X, and it has a backup replicate with port 6, instance Y.

Initially port 5 is up, so packets for this replica go to port 5 with instance X.

At some point later port 5 goes down, while port 6 is up.  It seems pretty clear from what you have written already that packets for this replica should now go to port 6, with instance Y.

What if port 6 remains up, and later port 5 comes back up again?  Do you expect/require that packets for this replica go back to port 5, instance X?  Or should the choice of replica only be re-evaluated if the port for where replicas are currently going, goes down?

I suppose a similar scenario applies to watch ports in P4Runtime, and I do not recall if it says explicitly what should happen in this scenario.

Also another minor nit: would it be just as clear to you to replace "becomes unavailable" with "goes down", and "first available backup replica" with "first backup replica for which its port is up"?  The terminology of a port being up or down is used when describing watch port behavior, but I don't recall whether the terms "available" and "unavailable" are used anywhere in the spec currently.

Reply:
Also, it might be nice to say explicitly what you expect to happen if the primary and all backup ports are down for a particular replica.

Do you expect no such copy to be made for that replica?

Use the last replica that was being used before it went down, if any of the ports were ever up, or the first one if none of the ports were ever up since the multicast group was last modified?

I realize the answer does not matter if there are no side effects as a result of performing after-multicast-replication processing on the packet before it is dropped (because the port is down), but there often _are_ side effects in after-multicast-replication processing, e.g. counter and meter updates.

Reply:
In the first situation that you mentioned Andy, we will go back to the original primary port whenever it comes back up. Will also make those changes as you've requested to specify up/down vs available/unavailable.

In the case where all the primary and all the backup ports are down for a particular replica, nothing is done to recover. We do not expect a copy to be made for that replica.

Reply:
I've updated the P4runtime Spec accordingly with the discussion. Please take a look @jafingerhut